File: src/frontend/renderer/api/personalSchema.ts
/**
 * API functions for personal schema management
 */

import {
  PersonalSchema,
  SchemaTemplate,
  EntityTypeDefinition,
  AddEntityTypeRequest,
  ApplyTemplateRequest,
} from '../../../shared/types/personalSchema';

const API_BASE = 'http://localhost:8001/api';

export async function getPersonalSchema(userId: string = 'default'): Promise<PersonalSchema> {
  const response = await fetch(`${API_BASE}/personal-schema?user_id=${userId}`);
  if (!response.ok) {
    throw new Error(`Failed to fetch personal schema: ${response.statusText}`);
  }
  return response.json();
}

export async function getSchemaTemplates(): Promise<SchemaTemplate[]> {
  const response = await fetch(`${API_BASE}/personal-schema/templates`);
  if (!response.ok) {
    throw new Error(`Failed to fetch schema templates: ${response.statusText}`);
  }
  return response.json();
}

export async function addEntityType(
  entityType: EntityTypeDefinition,
  userId: string = 'default'
): Promise<{ success: boolean; message: string; schema: PersonalSchema }> {
  const response = await fetch(`${API_BASE}/personal-schema/entity-types?user_id=${userId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ entity_type: entityType } as AddEntityTypeRequest),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to add entity type: ${error}`);
  }
  return response.json();
}

export async function removeEntityType(
  entityName: string,
  userId: string = 'default'
): Promise<{ success: boolean; message: string; schema: PersonalSchema }> {
  const response = await fetch(
    `${API_BASE}/personal-schema/entity-types/${entityName}?user_id=${userId}`,
    {
      method: 'DELETE',
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to remove entity type: ${error}`);
  }
  return response.json();
}

export async function applyTemplate(
  templateId: string,
  userId: string = 'default'
): Promise<{ success: boolean; message: string; schema: PersonalSchema }> {
  const response = await fetch(`${API_BASE}/personal-schema/apply-template?user_id=${userId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ template_id: templateId } as ApplyTemplateRequest),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to apply template: ${error}`);
  }
  return response.json();
}

export async function getEntityType(
  entityName: string,
  userId: string = 'default'
): Promise<EntityTypeDefinition> {
  const response = await fetch(
    `${API_BASE}/personal-schema/entity-types/${entityName}?user_id=${userId}`
  );

  if (!response.ok) {
    throw new Error(`Failed to fetch entity type: ${response.statusText}`);
  }
  return response.json();
}

export async function createSchema(schema: PersonalSchema): Promise<PersonalSchema> {
  const response = await fetch(`${API_BASE}/personal-schema`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(schema),
  });

  if (!response.ok) {
    throw new Error(`Failed to create schema: ${response.statusText}`);
  }
  return response.json();
}

// Export all functions as a namespace for convenience
export const personalSchemaApi = {
  getPersonalSchema,
  getSchemaTemplates,
  addEntityType,
  removeEntityType,
  applyTemplate,
  getEntityType,
  createSchema,
};

File: backend/api/routes/personal_schema.py
"""
API endpoints for personal schema management.
"""

from fastapi import APIRouter, HTTPException
from typing import List, Dict, Any
from pydantic import BaseModel
import logging

from core.personal_schema import (
    PersonalSchemaManager,
    EntityTypeDefinition,
    FieldDefinition,
    PersonalSchema,
    SchemaTemplate
)

router = APIRouter()
logger = logging.getLogger(__name__)

# Initialize schema manager
schema_manager = PersonalSchemaManager()

class AddEntityTypeRequest(BaseModel):
    """Request to add a new entity type."""
    entity_type: EntityTypeDefinition

class ApplyTemplateRequest(BaseModel):
    """Request to apply a template."""
    template_id: str

@router.get("/")
async def get_personal_schema(user_id: str = "default"):
    """Get user's personal schema configuration."""
    try:
        schema = schema_manager.get_schema(user_id)
        return schema.dict()
    except Exception as e:
        logger.error(f"Error getting personal schema: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/templates")
async def get_schema_templates():
    """Get all available schema templates."""
    try:
        templates = schema_manager.get_templates()
        return [t.dict() for t in templates]
    except Exception as e:
        logger.error(f"Error getting templates: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/entity-types")
async def add_entity_type(request: AddEntityTypeRequest, user_id: str = "default"):
    """Add a new entity type to user's schema."""
    try:
        schema = schema_manager.add_entity_type(user_id, request.entity_type)
        return {
            "success": True,
            "message": f"Added entity type '{request.entity_type.name}'",
            "schema": schema.dict()
        }
    except Exception as e:
        logger.error(f"Error adding entity type: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/entity-types/{entity_name}")
async def remove_entity_type(entity_name: str, user_id: str = "default"):
    """Remove an entity type from user's schema."""
    try:
        schema = schema_manager.remove_entity_type(user_id, entity_name)
        return {
            "success": True,
            "message": f"Removed entity type '{entity_name}'",
            "schema": schema.dict()
        }
    except Exception as e:
        logger.error(f"Error removing entity type: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/apply-template")
async def apply_template(request: ApplyTemplateRequest, user_id: str = "default"):
    """Apply a schema template."""
    try:
        schema = schema_manager.apply_template(user_id, request.template_id)
        return {
            "success": True,
            "message": f"Applied template '{request.template_id}'",
            "schema": schema.dict()
        }
    except Exception as e:
        logger.error(f"Error applying template: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/entity-types/{entity_name}")
async def get_entity_type(entity_name: str, user_id: str = "default"):
    """Get details of a specific entity type."""
    try:
        schema = schema_manager.get_schema(user_id)
        if entity_name not in schema.entity_types:
            raise HTTPException(status_code=404, detail=f"Entity type '{entity_name}' not found")
        
        return schema.entity_types[entity_name].dict()
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting entity type: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))